### ISSUE TITLE
Remote form without inputs sends corrupt multipart request

### ISSUE THREAD

**Issue by martijnversluis** [hauleth/vanilla-ujs#28]
`vanilla-ujs` version: **1.2.0**

I'm running into issues with remote forms when running feature tests in Poltergeist. Apparently, in test environment authenticity tokens are not included in (remote) forms. Nevertheless, an empty `FormData` is passed to `LiteAjax` which generates an empty multipart request without closing boundary. The corrupt request raises an error in Rack, as shown in the RSpec output below.

I also investigated how `jquery-ujs` deals with remote forms. For forms with inputs it sends an URL encoded request and without inputs it is just a plain request without body. Besides that, when using `jquery-ujs` I could add a `type: :json` to the `button_to` tag which would send a empty JSON request.

I see a few solutions:

- post requests without data (and maybe all post request without files) should be `x-www-form-urlencoded`
- when the `FormData` object has no values, it should not be sent (just `xhr.send()` instead of `xhr.send(data)`)
- `vanilla-ujs` should support `data-type` on the remote form to able to switch to a JSON request

### RSpec output:

```
     1.2) Failure/Error: raise EOFError if eof
          
          EOFError:
            EOFError
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/rack-2.0.1/lib/rack/multipart/parser.rb:361:in `handle_empty_content!'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/rack-2.0.1/lib/rack/multipart/parser.rb:185:in `on_read'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/rack-2.0.1/lib/rack/multipart/parser.rb:72:in `block in parse'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/rack-2.0.1/lib/rack/multipart/parser.rb:70:in `loop'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/rack-2.0.1/lib/rack/multipart/parser.rb:70:in `parse'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/rack-2.0.1/lib/rack/multipart.rb:52:in `extract_multipart'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/rack-2.0.1/lib/rack/request.rb:472:in `parse_multipart'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/rack-2.0.1/lib/rack/request.rb:335:in `POST'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/rack-2.0.1/lib/rack/method_override.rb:39:in `method_override_param'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/rack-2.0.1/lib/rack/method_override.rb:27:in `method_override'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/rack-2.0.1/lib/rack/method_override.rb:15:in `call'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/rack-2.0.1/lib/rack/runtime.rb:22:in `call'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/activesupport-5.0.0.1/lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/actionpack-5.0.0.1/lib/action_dispatch/middleware/executor.rb:12:in `call'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/actionpack-5.0.0.1/lib/action_dispatch/middleware/static.rb:136:in `call'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/rack-2.0.1/lib/rack/sendfile.rb:111:in `call'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/railties-5.0.0.1/lib/rails/engine.rb:522:in `call'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/rack-2.0.1/lib/rack/urlmap.rb:68:in `block in call'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/rack-2.0.1/lib/rack/urlmap.rb:53:in `each'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/rack-2.0.1/lib/rack/urlmap.rb:53:in `call'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/capybara-2.8.1/lib/capybara/server.rb:43:in `call'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/puma-3.6.0/lib/puma/server.rb:578:in `handle_request'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/puma-3.6.0/lib/puma/server.rb:415:in `process_client'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/puma-3.6.0/lib/puma/server.rb:275:in `block in run'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/puma-3.6.0/lib/puma/thread_pool.rb:116:in `call'
          # /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/puma-3.6.0/lib/puma/thread_pool.rb:116:in `block in spawn_thread'
          # 
          #   Showing full backtrace because every line was filtered out.
          #   See docs for RSpec::Configuration#backtrace_exclusion_patterns and
          #   RSpec::Configuration#backtrace_inclusion_patterns for more information.
          # ------------------
          # --- Caused by: ---
          # Capybara::CapybaraError:
          #   Your application server raised an error - It has been raised in your test code because Capybara.raise_server_errors == true
          #   /usr/local/var/rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/capybara-2.8.1/lib/capybara/session.rb:126:in `raise_server_error!'
```

### Comparison between `vanilla-ujs` and `jquery-ujs`

| Library       | Form inputs | `Content-Type` request header                                                       | Request body                             |
| ------------- | :---------: | ----------------------------------------------------------------------------------- | ---------------------------------------- |
| `vanilla-ujs` | no          | `Content-Type:multipart/form-data; boundary=----WebKitFormBoundarytlflAqgCb2rxgZF7` | ![image][vanilla_ujs_form_without_input] |
| `vanilla-ujs` | yes         | `Content-Type:multipart/form-data; boundary=----WebKitFormBoundarylncJBPQwE3m4lhEm` | ![image][vanilla_ujs_form_with_input]    |
| `jquery-ujs`  | no          | -none-                                                                              | -none-                                   |
| `jquery-ujs`  | yes         | `Content-Type:application/x-www-form-urlencoded; charset=UTF-8`                     | ![image][jquery_ujs_form_with_input]     |

[vanilla_ujs_form_without_input]: https://cloud.githubusercontent.com/assets/1219904/20135456/0e2042b8-a670-11e6-9d62-c95b3642b76e.png
[vanilla_ujs_form_with_input]: https://cloud.githubusercontent.com/assets/1219904/20135487/33b9f2ee-a670-11e6-881b-d9bb653ab645.png
[jquery_ujs_form_with_input]: https://cloud.githubusercontent.com/assets/1219904/20134717/be7b41e8-a66c-11e6-9cbe-3e8173836738.png

**Comment by hauleth** [comment]
Thanks for great writing. I will try to fix it ASAP.

**Comment by martijnversluis** [comment]
@hauleth I managed to write a fix, I issued a pull request here: #29.

A minor note:

`form.js` contains the line:

```js
data[CSRF.param()] = CSRF.token();
```

This structure does not always work (at least not in Chrome). I believe the recommended syntax is:

```js
data.append(CSRF.param(), CSRF.token());
```

On the other hand, isn't the input for the `authenticity_token` already included in the form automatically?
